### ISSUE TITLE
HasManyThrough, final Model id overwritten by intermediate model id if columns match

### ISSUE THREAD

**Issue by hailwood** [laravel/framework#23548]
- Laravel Version: 5.5.3
- PHP Version: 7.2
- Database Driver & Version: PostgresQL 9.6 although I believe this is agnostic

### Description:
Let's take these three models. We're using a custom intermediate model here rather than just a belongsToMany for reasons unrelated to this issue

User: id
Block: id, user_id (the owner)
BlockContractor: id, user_id (the contractor id), block_id

If we say that `User` hasMany `Block` through `BlockContractor` this makes sense.
However the query that gets executed ends up being `select blocks.*, block_contractor.user_id from ...` which means **the `user_id` from `block_contractor` overrides the actual `user_id` from `block` in the Block model we get back!**

An easy solution to this would be to allow us to alias the key e.g. 

```php
  public function contractedBlocks(): HasManyThrough
  {
    return $this->hasManyThrough(
      Block::class,
      ContractorBlockLink::class,
      'user_id AS contractor_id',
      'id',
      'id',
      'block_id'
    );
  }
```

However this almost works but ends up producing (pay attention to the last line)

```sql
select
  "blocks".*,
  "block_contractor"."user_id" as "contractor_id"
from "blocks"
  inner join "block_contractor" on "block_contractor"."block_id" = "blocks"."id"
where "block_contractor"."user_id" as "contractor_id" = 11
```

My solution right now is to just rename the column in block_contractor but that's not always going to be possible.

**Comment by sisve** [comment]
This is a known problem and discussed at https://github.com/laravel/ideas/issues/347.

**Comment by hailwood** [comment]
Ah thanks @sisve I'll join the conversation over there, but I'm going to also leave this open because of the fun we're having with the ideas repo.

**Comment by staudenmeir** [comment]
This will be fixed in Laravel 5.8 (#25812). Please close the issue.

**Comment by mf16** [comment]
For anyone else running into this issue before the 5.8 release, I just took the code changes that fixed the issue and created HasManyThroughFixed & a trait for using it.

```
<?php
// HasManyThroughFixedRelation.php

namespace App\Traits;

use Closure;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\HasManyThrough;
use App\Relations\HasManyThroughFixed;
use Illuminate\Database\Eloquent\Builder;

trait HasManyThroughFixedRelation
{
    public function hasManyThroughFixed($related, $through, $firstKey = null, $secondKey = null, $localKey = null, $secondLocalKey = null)
    {
        $through = new $through;

        $firstKey = $firstKey ?: $this->getForeignKey();

        $secondKey = $secondKey ?: $through->getForeignKey();

        return $this->newHasManyThroughFixed(
            $this->newRelatedInstance($related)->newQuery(),
            $this,
            $through,
            $firstKey,
            $secondKey,
            $localKey ?: $this->getKeyName(),
            $secondLocalKey ?: $through->getKeyName()
        );
    }

    protected function newHasManyThroughFixed(Builder $query, Model $farParent, Model $throughParent, $firstKey, $secondKey, $localKey, $secondLocalKey)
    {
        return new HasManyThroughFixed($query, $farParent, $throughParent, $firstKey, $secondKey, $localKey, $secondLocalKey);
    }
}
```
```
<?php
// HasManyThroughFixed.php

namespace App\Relations;

use Illuminate\Database\Eloquent\Relations\HasManyThrough;
use Illuminate\Database\Eloquent\Collection;

class HasManyThroughFixed extends HasManyThrough
{
    // Override 2 necessary methods to keep sane ids
    protected function buildDictionary(Collection $results)
    {
        $dictionary = [];

        // First we will create a dictionary of models keyed by the foreign key of the
        // relationship as this will allow us to quickly access all of the related
        // models without having to do nested looping which will be quite slow.
        foreach ($results as $result) {
            $dictionary[$result->_first_key][] = $result;
        }

        return $dictionary;
    }

    protected function shouldSelect(array $columns = ['*'])
    {
        if ($columns == ['*']) {
            $columns = [$this->related->getTable().'.*'];
        }

        return array_merge($columns, [$this->getQualifiedFirstKeyName().' as _first_key']);
    }
}
```
So now, just use that the `HasManyThroughFixedRelation` trait in your model, and use HasManyThroughFixed() just like you would HasManyThrough().
